### PR TITLE
Add rule S4123 ('no-invalid-await') for TS

### LINKED ISSUE #1516: Add rule S4123 ('no-invalid-await' in tslint) for TS

### PR DESCRIPTION
**Pull request by yassin-kammoun-sonarsource**
Fixes #1516

### CODE REVIEW

File: eslint-bridge/src/rules/no-invalid-await.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+};
+
+function getTypeFromTreeNode(node: estree.Node, services: RequiredParserServices) {
+  const checker = services.program.getTypeChecker();
+  return checker.getTypeAtLocation(services.esTreeNodeToTSNodeMap.get(node as TSESTree.Node));
+}
+
+function hasThenMethod(type: any, ts: any) {
+  const thenProperty = type.getProperty("then");
+  return Boolean(thenProperty && thenProperty.flags & ts.SymbolFlags.Method);
```
Comment (christophe-zurn-sonarsource) [inline eslint-bridge/src/rules/no-invalid-await.ts]: For readability, we could add parentheses around the bitewise operator: `Boolean(thenProperty && (thenProperty.flags & ts.SymbolFlags.Method));`

File: eslint-bridge/tests/RuleTesterTs.ts [hunk 2]
```diff
@@ -1,77 +1,78 @@
-/*
```
[the remark this answers is not in the document]
Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/tests/RuleTesterTs.ts]: Sorry about that. It's probably a difference in the line endings. There is actually one little change (see next comment).

File: eslint-bridge/tests/RuleTesterTs.ts [hunk 3]
```diff
@@ -1,77 +1,78 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import { RuleTester, Rule } from "eslint";
+import * as path from "path";
+
+const parser = path.resolve(`${__dirname}/../node_modules/@typescript-eslint/parser`);
+
+const parserOptions = {
+  ecmaVersion: 2018,
+  sourceType: "module",
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/tests/RuleTesterTs.ts]: Was added.

Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/tests/RuleTesterTs.ts]: I will fix that.

File: eslint-bridge/src/rules/no-invalid-await.ts (hunk 1 above)
Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/rules/no-invalid-await.ts]: It turned out it's not possible because the prettier removes those additional parentheses.

Comment (christophe-zurn-sonarsource) [reply eslint-bridge/src/rules/no-invalid-await.ts]: Ok. Nervermind then :)

**Pull request merged**